### PR TITLE
[XLA] Replace GCC vector extension with portable Intel SIMD

### PR DESCRIPTION
**Pull request by rongjiecomputer**
MSVC does not support GCC vector extension, so replace it with Intel SIMD library. Eigen also uses Intel SIMD library to implement vector functions.

MSVC does not have `__SSE4_1__` macro, although it does define `__AVX__` when building with `/arch:AVX`. Since Eigen enables SSE4.1 when `__AVX__` is defined ([`Eigen/Core`](https://bitbucket.org/eigen/eigen/src/034b6c3e101792a3cc3ccabd9bfaddcabe85bb58/Eigen/Core?at=default&fileviewer=file-view-default#Core-156)), we just follow Eigen.

#15213

### CODE REVIEW

**Comment by yifeif** [comment]
@sanjoy might taking a look? Thanks!

**Review by sanjoy — CHANGES_REQUESTED** [review]
Why do we need the new `TF_XLA_HAS_*` defines?

File: tensorflow/compiler/xla/service/cpu/cpu_runtime_sse4_1.h [hunk 1]
```diff
@@ -24,22 +24,33 @@ limitations under the License.
...
 
+// MSVC does not have __SSE4_1__ macro. Eigen enables EIGEN_VECTORIZE_SSE4_1
+// when __AVX__ is defined, we should do the same.
+#if defined(__SSE4_1__) || defined(__AVX__)
+#include <smmintrin.h>
+#define TF_XLA_HAS_SSE4_1 1
+#else
+#define TF_XLA_HAS_SSE4_1 0
+#endif
+
```
Comment (rongjiecomputer) [inline tensorflow/compiler/xla/service/cpu/cpu_runtime_sse4_1.h]: > Why do we need the new TF_XLA_HAS_* defines?

I use `__AVX__` to detect SSE4.1 for MSVC, so the macro to detect SSE4.1 becomes `defined(__SSE4_1__) || defined(__AVX__)`. Putting `defined(__SSE4_1__) || defined(__AVX__)` in every place might confuse reader since I only document the MSVC case in this place only. `TF_XLA_HAS_SSE4_1` provides clearer meaning.

tl;dr: For readability.

File: tensorflow/compiler/xla/service/cpu/simple_orc_jit.cc [hunk 2]
```diff
@@ -102,21 +102,9 @@ llvm::StringRef GetHostCpuName() {
...
-#endif
-#ifdef __ARM_NEON__
-  intrinsics.neon_intrinsics = true;
-#else
-  intrinsics.neon_intrinsics = false;
-#endif
+  intrinsics.sse_intrinsics = TF_XLA_HAS_SSE4_1;
+  intrinsics.avx_intrinsics = TF_XLA_HAS_AVX;
+  intrinsics.neon_intrinsics = TF_XLA_HAS_NEON;
   return intrinsics;
```
Comment (rongjiecomputer) [inline tensorflow/compiler/xla/service/cpu/simple_orc_jit.cc]: Also, with `TF_XLA_HAS_*` macro, this part becomes much shorter.

File: tensorflow/compiler/xla/service/cpu/cpu_runtime_sse4_1.h (hunk 1 above)
Comment (sanjoy) [reply tensorflow/compiler/xla/service/cpu/cpu_runtime_sse4_1.h]: Ok, that makes sense.  I had missed the fact that MSVC does not define  `__SSE4_1__`.

File: tensorflow/compiler/xla/service/cpu/cpu_runtime_sse4_1.h [hunk 3]
```diff
@@ -24,22 +24,33 @@ limitations under the License.
 
 #include "tensorflow/core/platform/macros.h"
 
+// MSVC does not have __SSE4_1__ macro. Eigen enables EIGEN_VECTORIZE_SSE4_1
+// when __AVX__ is defined, we should do the same.
+#if defined(__SSE4_1__) || defined(__AVX__)
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/cpu/cpu_runtime_sse4_1.h]: I'd suggest putting all of these definitions in a single header; and have this `defined(__SSE4_1__) || defined(__AVX__)` logic only for MSVC.  I don't think the latter bit makes a functional difference since clang defines `__SSE4_1__` when building with `-mavx`, but it will make the motivation clearer.

I'd also avoid `#define TF_XLA_HAS_SSE4_1 1` and just have `#define TF_XLA_HAS_SSE4_1`.  Even though what you have makes `GetAvailableIntrinsics` shorter, it allows us to mistakenly write `#ifdef TF_XLA_HAS_SSE4_1`.

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/cpu/cpu_runtime_sse4_1.h (hunk 3 above)
Comment (rongjiecomputer) [reply tensorflow/compiler/xla/service/cpu/cpu_runtime_sse4_1.h]: > and have this defined(__SSE4_1__) || defined(__AVX__) logic only for MSVC

Done. Becomes `defined(__SSE4_1__) || (defined(_MSC_VER) && defined(__AVX__))`.

> I'd also avoid #define TF_XLA_HAS_SSE4_1 1 and just have #define TF_XLA_HAS_SSE4_1. Even though what you have makes GetAvailableIntrinsics shorter, it allows us to mistakenly write #ifdef TF_XLA_HAS_SSE4_1.

Done. I did mistakenly write `#ifdef` while working on the old patch, so I can understand your concern.

> I'd suggest putting all of these definitions in a single header

I have eventually decided to not splitting the definitions out into a new header file. That requires adding a new `cc_library` and make three `cc_library(name = "cpu_runtime_*")` to depend on it due to Bazel's strict header dependency. Seems over-complicated.

**Comment by drpngx** [comment]
@rongjiecomputer did you address @sanjoy comments?

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by rongjiecomputer** [comment]
@drpngx 

> @rongjiecomputer did you address @sanjoy comments?

Only partially. @sanjoy left these two comments.

1. Avoid `#define TF_XLA_HAS_SSE4_1 1` and just have `#define TF_XLA_HAS_SSE4_1`: I have done this part.

2. Putting all of these definitions (`TF_XLA_HAS_*`) in a single header: I gave the explanation why I did not follow this comment.

> I have eventually decided to not splitting the definitions out into a new header file. That requires adding a new cc_library and make three cc_library(name = "cpu_runtime_*") to depend on it due to Bazel's strict header dependency. Seems over-complicated.

**Comment by drpngx** [comment]
SG, thanks! Let's what what @sanjoy thinks.

**Review by sanjoy — APPROVED** [review]
LGTM!
